spec/lib/hubspot/contact_spec.rb

```diff
@@ -52,34 +52,54 @@
   end
 
   describe '.createOrUpdate' do
-    cassette 'contact_create_or_update'
-    let(:params){{}}
-    subject{ Hubspot::Contact.createOrUpdate(email, params) }
-    context 'with a new email' do
-      let(:email){ "newcontact#{Time.now.to_i}@hsgem.com" }
-      it{ should be_an_instance_of Hubspot::Contact }
-      its(:email){ should match /newcontact.*@hsgem.com/ } # Due to VCR the email may not match exactly
-
-      context 'and some params' do
-        cassette 'contact_create_or_update_with_params'
-        let(:email){ "newcontact_x_#{Time.now.to_i}@hsgem.com" }
-        let(:params){ {firstname: 'Hugh', lastname: 'Jackman' } }
-        its(['firstname']){ should == 'Hugh'}
-        its(['lastname']){ should == 'Jackman'}
+    context "when the contact already exists" do
+      it "updates the contact" do
+        VCR.use_cassette("contacts/update_contact", record: :none) do
+          existing_contact = Hubspot::Contact.create!("contact@example.com")
+          new_email = "new_email@example.com"
+          params = { email: new_email }
+
+          contact = Hubspot::Contact.createOrUpdate(
+            existing_contact.email,
+            params
+          )
+
+          assert_requested :post, hubspot_api_url("/contacts/v1/contact/createOrUpdate/email/#{existing_contact.email}?hapikey=demo")
+          expect(contact).to be_a(Hubspot::Contact)
+          expect(contact.email).to eq(new_email)
+
+          contact.destroy!
```

Comment:
Does this happen a lot? Should this be abstracted out similar to `DatabaseCleaner`?

Reply:
@tabfugnic  
I'm glad you mentioned this because this is one area that continues to surprise me about VCR. While destroying records doesn't happen very often, it will become more prevalent and
I haven't found any resources that discuss the setup and cleanup process that's required with VCR. Most examples use `GET` requests that rely on data existing.

In my mind, destroying the created record isn't quite as straightforward as using `DatabaseCleaner` because:
1) It's probably best to delete just the created record vs delete all contacts
2) each request made in the test is captured in the VCR cassette. Using this test as an example, the requests to remove "all" contacts would be included in each cassette.

If I create and delete the records outside of the `use_cassette` block, the tests will issue real network calls each time the test is run.
<details>
<summary>Create and Delete outside of VCR</summary>

```
it "updates the contact" do
  existing_contact = Hubspot::Contact.create!("contact@example.com")

  VCR.use_cassette("contacts/update_contact", record: :none) do
    contact = Hubspot::Contact.createOrUpdate(email, params)
   ... epectations ....
  end

  contact.destroy!
end
```
</details>
</br>
With that in mind, I'm not sure how to treat the "cleanup" of these records without having to do manual setup and cleanup inline. Do you have any suggestions?

Reply:
That is fascinating! Thank you for the very comprehensive response.

I wonder if there is some trickery that you could do with `#insert_cassette` and `#eject_cassette` so you have a bit more control over when these happen...

Reply:
I had one other thought. The records don't technically have to be deleted. The important part is ensuring the necessary test setup is correct. I could extract a function that executes the `create a contact` call and only raises if the response is something unexpected, like a 500.

Pros: Data doesn't have to be destroyed at the end of each test that creates data
Cons: Given when the cassette is recorded, the `post` recording will likely include the "record already exists" response.

Reply:
Since this is using VCR, the destroy only happens the first time the cassette is created or when the cassette is re-generated, not every time the test runs so it's not really a big deal in terms of improving test performance and DRYing out tests like DatabaseCleaner is typically used for.

Additionally, since this is an API call to a remote system not under the control of the test infrastructure, cleaning the records would require custom functionality to understand which records were created including parsing the identifier (or similar method for identifying a unique object) and logic specific to each object for deleting that object from the remote system. Currently the tests all utilize the `demo` account which is provided by HubSpot but it's fairly easy to create a developer account where you can get a unique API key so your created objects will be unique to that account.